core/scmi-server/scmi_server.c

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019-2022, Linaro Limited
+ */
+
+#include <arch_main.h>
+#include <config.h>
+#include <initcall.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <scmi/scmi_server.h>
+#include <optee_scmi.h>
+
+/*
+ * OP-TEE helper function exported to SCP-firmware
+ */
+uintptr_t smt_phys_to_virt(uintptr_t pa, size_t sz, bool shmem_is_secure)
+{
+	if (shmem_is_secure)
+		return (uintptr_t)phys_to_virt(pa, MEM_AREA_IO_SEC, sz);
+	else
+		return (uintptr_t)phys_to_virt(pa, MEM_AREA_IO_NSEC, sz);
+}
+
+/*
+ * SCMI server APIs exported to OP-TEE core
+ */
+int scmi_server_get_channels_count(void)
+{
+	return scmi_get_devices_count();
+}
+
+TEE_Result scmi_server_get_channel(unsigned int channel_id, int *handle)
+{
+	int fwk_id = 0;
+
+	fwk_id = scmi_get_device(channel_id);
+	if (fwk_id < 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (handle)
```

Comment:
In the case `handle == NULL` I think it would be better to return `TEE_ERROR_BAD_PARAMETERS`. :)
Maybe:
```c
if (handle == NULL)
     return TEE_ERROR_BAD_PARAMETERS;

*handle = fwk_id;
```

Reply:
> Actaully there is a call to `scmi_server_get_channel()` from the SCMI PTA that only checks the channel ID is valid but does not retrieve the SCP-fmw related handle, [here in current patch](https://github.com/OP-TEE/optee_os/pull/5627/commits/8b96706fe7a5d7ed0033304fa3a5c5aec6997985#diff-b02cf9d13860a7b8eb818540e5776a98cba205eecc11519571d540cac8b5fa61R193-R196).

That‘s ok